chips/stm32f429zi/src/ethernet/ethernet.rs

```diff
@@ -0,0 +1,2288 @@
+    ///
+    /// # Errors
+    ///
+    /// + [Err]\([ErrorCode::OFF]\): the reception of the Ethernet peripheral is off
+    ///
+    /// # Panics
+    ///
+    /// This method panics if no receive packet buffer has been set
+    pub fn receive_packet(&self) -> Result<(), ErrorCode> {
+        // Check if DMA and MAC core are enabled
+        if !self.is_reception_enabled() {
+            return Err(ErrorCode::OFF);
+        }
+
+        let received_packet = self.received_packet.take().unwrap();
+        self.receive_descriptor
+            .set_buffer1_address(received_packet.as_ptr() as u32);
+        self.receive_descriptor
+            .set_buffer1_size(received_packet.len())?;
+        self.receive_descriptor.set_buffer2_size(0)?;
+        self.received_packet.put(Some(received_packet));
+
+        // Acquire the receive descriptor
+        self.receive_descriptor.acquire();
+
+        // Wait 4 CPU cycles until everything is written to the RAM
+        for _ in 0..4 {
+            nop();
+        }
+
+        self.dma_receive_poll_demand();
+
+        Ok(())
+    }
+}
+
+impl Configure for Ethernet<'_> {
+    fn init(&self) -> Result<(), ErrorCode> {
+        self.clocks.enable();
+        self.init_transmit_descriptors();
+        self.init_receive_descriptors();
+        self.init_dma()?;
+        self.init_mac()?;
+
+        Ok(())
+    }
+
+    fn set_operation_mode(&self, operation_mode: OperationMode) -> Result<(), ErrorCode> {
+        if self.is_mac_enabled() {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.internal_set_operation_mode(operation_mode);
+
+        Ok(())
+    }
+
+    fn get_operation_mode(&self) -> OperationMode {
+        self.internal_get_operation_mode()
+    }
+
+    fn set_speed(&self, speed: EthernetSpeed) -> Result<(), ErrorCode> {
+        if self.is_mac_enabled() {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.set_ethernet_speed(speed);
+
+        Ok(())
+    }
+
+    fn get_speed(&self) -> EthernetSpeed {
+        self.get_ethernet_speed()
+    }
+
+    fn set_loopback_mode(&self, enable: bool) -> Result<(), ErrorCode> {
+        match enable {
+            false => self.disable_loopback_mode(),
+            true => self.enable_loopback_mode(),
+        };
+
+        Ok(())
+    }
+
+    fn is_loopback_mode_enabled(&self) -> bool {
+        self.internal_is_loopback_mode_enabled()
+    }
+
+    fn set_mac_address(&self, mac_address: MacAddress) -> Result<(), ErrorCode> {
+        self.set_mac_address0(mac_address);
+
+        Ok(())
+    }
+
+    fn get_mac_address(&self) -> MacAddress {
+        self.get_mac_address0()
+    }
+
+    // TODO: Remove this once Transmit trait is implemented
+    fn start_transmit(&self) -> Result<(), ErrorCode> {
+        self.enable_transmitter()
+    }
+
+    // TODO: Remove this once Receive trait is implemented
+    fn start_receive(&self) -> Result<(), ErrorCode> {
+        self.enable_receiver()
+    }
```

Comment:
I believe these functions are no longer relevant? In general, a receive API will likely be asynchronous, with the MAC maintaining a (shared) buffer poll which it can pass to a registered client on demand.

Reply:
The story is that when I started working on the Ethernet, I was not aware of the `tock-ethernet` branch, so I started to create my own Ethernet HIL. Since Ethernet is another mean for transmitting and receiving data, I mimicked the UART HIL. Now, all traits are removed from the HIL.